### PR TITLE
close a bunch of issues

### LINKED ISSUE #14: Ability to get the resource update article / changelog
Would be great if I can get the update article of my resources, so then I can make a discord bot that will announce the update + the change log at the same time without me writing it by my own at my discord channel

### LINKED ISSUE #22: Add icon URL to getResource option
The `getResource` option should have an additional value, where the Icon URL of the resource is provided.
This could/should perhaps return the URL for the default resource icon, if none is set.

### LINKED ISSUE #24: Suggestion: add a search option
Add search option to get an array of the users that have that name

### LINKED ISSUE #30: Identities returning wrong keys
Discord key is now facebook
yahoo key returned twitter

https://api.spigotmc.org/simple/0.1/index.php?action=getAuthor&id=106

### LINKED ISSUE #31: Add supported versions
It would be useful, if the `getResource` action would also return the "Tested Minecraft versions" output from a resource (Maybe either as a JSON array or as a comma-separated String)

It could perhaps be called something like `mc_versions`.

### LINKED ISSUE #32: Endpoint to get resource list
It would be cool to have an endpoint similar to getResourcesByAuthor (maybe getResources) which returned a similar list as the author endpoint with all available resources (like the normal [resource listing](https://www.spigotmc.org/resources/) on the page does). It should include a parameter to specify the ordering as well as filtering for the category that gets displayed. (Edited to include all resources and have categories/sorting via parameter)

This would need some kind of pagination though as the list of resources would obviously be a lot longer than the author list and should probably not be all returned in one array.

### LINKED ISSUE #34: Expose resource description.
bb code pls.

### LINKED ISSUE #35: API Missing Important Information
When you visit a plugin's overview page, you are able to see the "total" amount of reviews in the small selection tab (https://prnt.sc/uwx8tk) and the "unique" amount of reviews in the information box (https://prnt.sc/uwx9cu). However, the API only returns the "total" amount of reviews only, which after a quick conversation with a SpigotMC resource staff member, we established this doesn't make sense, and I was instructed to open an issue about this here. If possible, please add the ability to retrieve this data.

### LINKED ISSUE #36: Identities privacy settings are discarded
In the privacy settings on the forums there is an option to choose who can view your identities. This option is discarded when accessing information through this API as one can perform the ```getAuthor``` action and view user's identities that one can't view on the forums.

Identities should be exposed via this API only if a user has set them visible to the public.

### PR DESCRIPTION
**Pull request by simpleauthority**
Intends to:
- close #14 (done)
- close #22 (done)
- close #24 (done)
- close #30 (done, i hope)
- close #31 (done)
- close #32 (done)
- close #34 (done)
- close #35 (done)
- close #36 (done)

### CODE REVIEW

**Comment by simpleauthority** [comment]
Replaces #37 because I butchered the other one.

**Comment by simpleauthority** [comment]
Going to do some work on this soon. Have had too much school stuff going on recently.

**Comment by simpleauthority** [comment]
#31 done, but it's done based on assumption of spigot's database schema matching what I have and how they have set it up...moving on, hope it works. Otherwise file more issues ;p

**Comment by simpleauthority** [comment]
#22 and #24 should both be good to go...trucking along

**Comment by simpleauthority** [comment]
#35 and #14 should be ok

going to do #32 later, my brain needs a break

**Comment by simpleauthority** [comment]
@md-5 This is ready for review. For your reference:

**ADDITIONS**
- `getResourceUpdate` and `getResourceUpdates` routes have been added. For the former, it is requesting a specific update by its update id. For the latter, it is finding all updates for a given resource by the resource id. This addresses #14.

- Resources and authors now contain the respective icon URL if one exists. Otherwise, it will return the default icon. This makes use of the `PUBLIC_PATH` config setting. We had this before, but did not want to keep it because it was downloading every image and base64 encoding it. This simply provides a link, and the user will have to do whatever they want with it at that point. This addresses #22.

- A `findUser` route was added which accepts a `name` parameter and finds a user by exact username match. This is not a general search feature, so it does not fully honor the issue, but I did not want to go too far and add too much to this route. Maybe I can add more later. Depends on demand. This addresses #24.

- A `listResources` route which accepts `category` and `page` arguments. The `category` is numeric and refers to the ID of the category (as opposed to name), and the `page` is numeric defaulting to `1`. Response is truncated to 10 results per page and this is not user modifiable. Can be decreased or increased as required. If no category is specified, resources will simply be returned in ascending order by resource id. To accommodate the numeric category, a second route `listResourceCategories` was made available. This route will show available category IDs with title and description. This addresses #32.

**CHANGES**
- A docker environment was added so anyone can run the API and help code. They just need a XenForo w/ XenForo Resource Manager instance to point it at.

- Hopefully identities are fixed this time around, but like before it's still a guessing game. If so, this addresses #30.

- The native and supported minecraft versions are now shown on the resource route(s). This addresses #31.

- The resource description has been exposed on the resource route(s). This addresses #34.

- A discrepancy in what the API reports as 'reviews' was identified in that what we show is the total amount of reviews and not unique amount of reviews across all downloads. The API now shows both unique and total review counts. This addresses #35.

- The API now checks whether or not a user actually _wants_ to share their identities before putting them in the response. This addresses #36.

And that should be it!

**EXTRA**
The DB fields needed have slightly changed for resource/author and there are some extra queries to look at. If you have any questions or need clarification please just reach out.

Thanks,
simple

**Comment by md-5** [comment]
Nice work

File: src/object/Resource.php [hunk 1]
```diff
@@ -0,0 +1,83 @@
...
+                case 'native_mc_version':
+                    $this->native_minecraft_version = substr_replace($field['actual_field_value'], ".", 1, 0);
+                    break;
+                case 'mc_versions':
+                    // This is going to be gross, I promise. Idk what this format is :^)
+                    // https://github.com/SpigotMC/XenforoResourceManagerAPI/issues/31
+                    $weird_format = $field['actual_field_value'];
+                    
+                    $extracted = array();
+                    preg_match("/^.+{(.+)}$/", $weird_format, $extracted);
```
[the remark this answers is not in the document]
Comment (simpleauthority) [reply src/object/Resource.php:38]: LOL 🤦‍♂️You're right...

**Comment by md-5** [comment]
Looks good as far as my php skills go.
Trying to find an extra pair of eyes or two would probably be helpful, but otherwise I'm ok with merging

One thing probably worth double checking though, does this **need** php 8? 
We currently run 7.3 as that's the supported debian version https://packages.debian.org/buster/php

**Review by marc-maurer — APPROVED** [review]
Looks good to me.
Great work!

**Comment by simpleauthority** [comment]
> 
> 
> Looks good to me.
> Great work!

Thank you for reviewing Marc, I appreciate it!

> 
> 
> Looks good as far as my php skills go.
> Trying to find an extra pair of eyes or two would probably be helpful, but otherwise I'm ok with merging
> 
> One thing probably worth double checking though, does this **need** php 8?
> We currently run 7.3 as that's the supported debian version https://packages.debian.org/buster/php

Does not need php 8, nope. I just updated the docker stuff to use 7.3 to match Spigot's env. I tested every route and all works as expected!

I asked Marc to review and I'm looking for one more person, or you can merge if you feel okay with it already.

Thanks both!

**Comment by md-5** [comment]
Excellent stuff. Thanks very much Marc, likewise I really appreciate it. 

All good with this from me and can merge if you wish. Unfortunately can't deploy right now as otherwise preoccupied, but I will make sure it is deployed ASAP.

Thanks again for your hard work Simple, also really appreciate it.

**Comment by simpleauthority** [comment]
Ready to merge!

Thanks a lot md! Good luck with the 1.17 update 😎

**Comment by simpleauthority** [comment]
- added page to `getResourceUpdates` as I forgot to do so. Now in case a resource is particularly prolific in updates, the request to the API won't murder the database
- made some changes for consistency in SQL query builder methods
- squashed commits

**Comment by simpleauthority** [comment]
Okay, now it's really ready to go. Just had some itches to scratch there. Not touching anymore 👍🏻

**Review by marc-maurer — APPROVED** [review]

**Comment by md-5** [comment]
Merged and will work on deploying

**Pull request merged**